charmtools/build/builder.py

```diff
@@ -638,28 +642,53 @@ def _validate_charm_repo(self):
             log.warn(msg)
 
     def __call__(self):
-        self.find_or_create_repo()
-
         log.debug(json.dumps(
             self.status(), indent=2, sort_keys=True, default=str))
         self.validate()
+        self.find_or_create_target()
         self.generate()
 
     def inspect(self):
         self.charm = path(self.charm).abspath()
         self._check_path(self.charm)
         inspector.inspect(self.charm, force_styling=self.force_color)
 
-    def normalize_outputdir(self):
-        od = path(self.charm).abspath()
-        repo = os.environ.get('JUJU_REPOSITORY')
-        if repo:
-            repo = path(repo)
-            if repo.exists():
-                od = repo
-        elif ":" in od:
-            od = od.basename
-        self.output_dir = od
+    def normalize_build_dir(self):
+        charm_build_dir = os.environ.get('CHARM_BUILD_DIR')
+        juju_repo_dir = os.environ.get('JUJU_REPOSITORY')
+        if not self.build_dir:
+            if charm_build_dir:
+                self.build_dir = path(charm_build_dir)
+            elif juju_repo_dir:
+                series = self.series or 'builds'
+                self.build_dir = path(juju_repo_dir) / series
+            else:
+                series = self.series or 'builds'
+                log.warn('Build dir not specified via command-line or '
+                         'environment; defaulting to /tmp/charm-builds')
+                self.build_dir = path('/tmp/charm-builds')
```

Comment:
I'm vaguely concerned that `/tmp/charm-builds` as a default could lead to conflicts in multi-user environments. But it's certainly not a big deal - `charm build` makes it pretty obvious that you can configure the build dir if needed.

Reply:
Yeah, I considered doing `$HOME/charm-builds` or something, but I didn't want to accidentally clobber anything with this change.  Also, it's better if the user is explicit, anyway, hence the warning.